workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectCreation/GameObjectCreatorFromMetadata.cs

```diff
@@ -0,0 +1,81 @@
+using System.Collections.Generic;
+using System.IO;
+using Generated.Improbable;
+using Improbable.Gdk.Core;
+using Improbable.Gdk.Core.GameObjectRepresentation;
+using UnityEngine;
+using Transform = Generated.Improbable.Transform.Transform;
+
+namespace Improbable.Gdk.Core.GameObjectRepresentation
+{
+    public class GameObjectCreatorFromMetadata : IEntityGameObjectCreator
+    {
+        private readonly Dictionary<string, GameObject> cachedPrefabs
+            = new Dictionary<string, GameObject>();
+
+        private readonly string workerType;
+
+        public GameObjectCreatorFromMetadata(string workerType)
+        {
+            this.workerType = workerType;
+        }
+
+        public GameObject CreateGameObjectForEntityAdded(SpatialOSEntity entity, WorkerSystem worker)
+        {
+            if (!entity.HasComponent<Metadata.Component>())
+            {
+                return null;
+            }
+
+            var prefabName = entity.GetComponent<Metadata.Component>().EntityType;
+            if (!entity.HasComponent<Transform.Component>())
+            {
+                cachedPrefabs[prefabName] = null;
+                return null;
+            }
+
+            var transform = entity.GetComponent<Transform.Component>();
+            var position = new Vector3(transform.Location.X, transform.Location.Y, transform.Location.Z) +
+                worker.Origin;
+            var rotation = new UnityEngine.Quaternion(transform.Rotation.X, transform.Rotation.Y,
+                transform.Rotation.Z, transform.Rotation.W);
+            var workerSpecificPath = Path.Combine("Prefabs", worker.WorkerType, prefabName);
+            var commonPath = Path.Combine("Prefabs", "Common", prefabName);
+
+            if (!cachedPrefabs.TryGetValue(prefabName, out var prefab))
+            {
+                prefab = Resources.Load<GameObject>(workerSpecificPath);
+                if (prefab == null)
+                {
+                    prefab = Resources.Load<GameObject>(commonPath);
+                }
+
+                if (prefab == null)
+                {
+                    worker.LogDispatcher.HandleLog(LogType.Log, new LogEvent(
+                        $"Prefab not found for SpatialOS Entity in either {workerSpecificPath} or {commonPath}," +
+                        "not going to associate a GameObject with it."));
+                }
+
+                cachedPrefabs[prefabName] = prefab;
+            }
+
+            if (prefab == null)
+            {
+                return null;
+            }
+
+            var gameObject = Object.Instantiate(prefab, position, rotation);
+            gameObject.name = $"{prefab.name}(SpatialOS: {entity.SpatialOSEntityId}";
```

Comment:
add the worker name to the name. I realized yesterday that it is extremely useful to know which game object belongs to which worker